### PR TITLE
feat(python): real pyenv support

### PR DESCRIPTION
**Pull request by bewing**
If `python` points at a pyenv shim, use `pyenv version-name` to
determine Python version and Venv information.

### Prerequisites

- [x] I have read and understood the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes/features)

### CODE REVIEW

**Comment by bewing** [comment]
This approach shaves about 100msec off python prompt generation on my WSL environment (since we never run `python` directly, and `pyenv` is 100% bash)

debug log for the segment:
```
python(true)         -  51 ms -   virt 3.9.9
2022/04/21 14:05:05 debug: Pwd
/home/bewing/python
2022/04/21 14:05:05 Pwd duration: 924ns, args:
2022/04/21 14:05:05 debug: Getenv

2022/04/21 14:05:05 Getenv duration: 1.322µs, args: VIRTUAL_ENV
2022/04/21 14:05:05 debug: Getenv

2022/04/21 14:05:05 Getenv duration: 1.161µs, args: CONDA_ENV_PATH
2022/04/21 14:05:05 debug: Getenv

2022/04/21 14:05:05 Getenv duration: 1.096µs, args: CONDA_DEFAULT_ENV
2022/04/21 14:05:05 debug: Pwd
/home/bewing/python
2022/04/21 14:05:05 Pwd duration: 915ns, args:
2022/04/21 14:05:05 PathSeparator duration: 44ns, args:
2022/04/21 14:05:05 debug: HasFiles
true
2022/04/21 14:05:05 HasFiles duration: 16.68µs, args: *.py
2022/04/21 14:05:05 debug: RunCommand
virt
2022/04/21 14:05:05 RunCommand duration: 31.667596ms, args: pyenv version-name
2022/04/21 14:05:05 TemplateCache duration: 763ns, args:
2022/04/21 14:05:05 TemplateCache duration: 481ns, args:

**Comment by bewing** [comment]
Need to refactor this code to make it mockable/testable.  Changes to `environment` probably needed

File: src/environment/shell.go [hunk 1]
```diff
@@ -500,22 +501,29 @@ func (env *ShellEnvironment) RunShellCommand(shell, command string) string {
 	return ""
 }
 
-func (env *ShellEnvironment) HasCommand(command string) bool {
+func (env *ShellEnvironment) HasCommandPath(command string) string {
```
Comment (JanDeDobbeleer) [inline src/environment/shell.go]: This implies a boolean return but isn't, can we rename this to `CommandPath`?

File: src/segments/python.go [hunk 2]
```diff
@@ -101,3 +95,41 @@ func (p *Python) canUseVenvName(name string) bool {
...
 	return true
 }
+
+func (p *Python) pyenvVersion() (string, error) {
+	// Use `pyenv root` instead of $PYENV_ROOT?
+	// Is our Python executable at $PYENV_ROOT/bin/python ?
+	// Should p.env expose command paths?
+	path := p.env.HasCommandPath("python")
+	// TODO:  pyenv-win has this at $PYENV_ROOT/pyenv-win/shims
+	if path == filepath.Join(p.env.Getenv("PYENV_ROOT"), "shims", "python") {
```
Comment (JanDeDobbeleer) [inline src/segments/python.go]: Reverse the statement so we can return early rather than indent this entirely?

**Review by JanDeDobbeleer — CHANGES_REQUESTED** [review]

File: src/segments/python.go (hunk 2 above)
Comment (bewing) [reply src/segments/python.go]: inverted.

File: src/environment/shell.go (hunk 1 above)
Comment (bewing) [reply src/environment/shell.go]: renamed

**Comment by bewing** [comment]
Think code and tests are good now.  Will look at docs next.  Do we want to feature gate this, or no?

**Comment by JanDeDobbeleer** [comment]
@bewing I wouldn't gate it, just ship. If you tested locally, we'll find out just how representative it was 😃

**Comment by bewing** [comment]
Bah, I thought I was testing locally, but I wasn't.  Not ready for review yet, but I'm not sure how to mark it as draft again

**Comment by bewing** [comment]
Found the error in my test case, fixed it.  Really really think it's ready now.

File: src/environment/shell.go [hunk 3]
```diff
@@ -401,6 +403,10 @@ func (env *ShellEnvironment) HasFolder(folder string) bool {
 	return hasFolder
 }
 
+func (env *ShellEnvironment) RealPath(path string) (string, error) {
```
Comment (JanDeDobbeleer) [inline src/environment/shell.go]: I would rename this to `ResolveSymlink` as that what this does.

**Review by JanDeDobbeleer — CHANGES_REQUESTED** [review]

File: src/environment/shell.go (hunk 3 above)
Comment (bewing) [reply src/environment/shell.go]: renamed

**Comment by JanDeDobbeleer** [comment]
Thanks for the changes! 🚀

**Pull request merged**